stacks-signer/src/chainstate.rs

```diff
@@ -0,0 +1,375 @@
+
+/// Captures this signer's current view of a sortition's miner.
+#[derive(PartialEq, Eq)]
+pub enum SortitionMinerStatus {
+    /// The signer thinks this sortition's miner is invalid, and hasn't signed any blocks for them.
+    InvalidatedBeforeFirstBlock,
+    /// The signer thinks this sortition's miner is invalid, but already signed one or more blocks for them.
+    InvalidatedAfterFirstBlock,
+    /// The signer thinks this sortition's miner is valid
+    Valid,
+}
+
+/// Captures the Stacks sortition related state for
+///  a successful sortition
+pub struct SortitionState {
+    /// The miner's pub key hash
+    pub miner_pkh: Hash160,
+    /// If known already, the public key which hashes to `miner_pkh`
+    pub miner_pubkey: Option<StacksPublicKey>,
+    /// the last burn block in this fork which had a sortition
+    pub prior_sortition: ConsensusHash,
+    /// the committed to parent tenure ID
+    pub parent_tenure_id: ConsensusHash,
+    /// this sortition's consensus hash
+    pub consensus_hash: ConsensusHash,
+    /// did the miner in this sortition do something
+    ///  to become invalidated as a miner?
+    pub invalidated: SortitionMinerStatus,
+}
+
+/// The signer's current view of the stacks chain's sortition
+///  state
+pub struct SortitionsView {
+    /// the prior successful sortition (this corresponds to the "prior" miner slot)
+    pub last_sortition: Option<SortitionState>,
+    /// the current successful sortition (this corresponds to the "current" miner slot)
+    pub cur_sortition: Option<SortitionState>,
+    /// is the view fresh?
+    pub fresh: bool,
+    /// the hash at which the sortitions view was last fetched
+    pub latest_consensus_hash: Option<ConsensusHash>,
+}
+
+impl TryFrom<SortitionInfo> for SortitionState {
+    type Error = ClientError;
+    fn try_from(value: SortitionInfo) -> Result<Self, Self::Error> {
+        Ok(Self {
+            miner_pkh: value
+                .miner_pk_hash160
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            miner_pubkey: None,
+            prior_sortition: value
+                .last_sortition_ch
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            consensus_hash: value.consensus_hash,
+            parent_tenure_id: value
+                .stacks_parent_ch
+                .ok_or_else(|| ClientError::UnexpectedSortitionInfo)?,
+            invalidated: SortitionMinerStatus::Valid,
+        })
+    }
+}
+
+enum ProposedBy<'a> {
+    LastSortition(&'a SortitionState),
+    CurrentSortition(&'a SortitionState),
+}
+
+impl<'a> ProposedBy<'a> {
+    pub fn state(&self) -> &SortitionState {
+        match self {
+            ProposedBy::LastSortition(ref x) => x,
+            ProposedBy::CurrentSortition(ref x) => x,
+        }
+    }
+}
+
+impl SortitionsView {
+    /// Initialize an empty sortitions view struct -- it will refresh() before
+    ///  checking any proposals.
```

Comment:
This kinda flies in the face of RAII.  Why not call `self.refresh()` here, and return a `Result<Self>`?

Reply:
But that's just the thing -- you're going to be calling `.refresh()` whenever you're checking a proposal.  Why not just have `refresh()` be a static method that returns an instance of this struct?

Reply:
Ah, I see. Yeah, I think that’s a good idea. The signer run loop will still want to cache this struct: it can be reused during a tenure, it just needs to be dropped whenever the runloop receives a burn chain event. But there’s no need to try to handle that at the level of this struct.